### PR TITLE
Adding workload API

### PR DESCRIPTION
**Pull request by huguesBouvier**
This PR adds:
- Workload API with full path the fetch the trust bundle
- Generic node to test the API
- Fully working API to fetch the trust bundle.

### CODE REVIEW

File: iot-edge-spiffe-agent/workload-api-server/src/unix_stream.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (huguesBouvier) [inline iot-edge-spiffe-agent/workload-api-server/src/unix_stream.rs:1]: Add this is a copy past from tonic and put a link to it

File: Cargo.toml [hunk 2]
```diff
@@ -18,8 +19,12 @@ members = [
 	"iot-edge-spiffe-agent/agentd",	
 	"iot-edge-spiffe-agent/config",	
 	"iot-edge-spiffe-agent/node-attestation",
-	"iot-edge-spiffe-agent/server-client",
+	"iot-edge-spiffe-agent/spiffe-server-client",
 ]
+exclude = [
```
Comment (nyanzebra) [inline Cargo.toml:24]: nit: maybe a small note somewhere in here that says we are excluding auto-generated code?

File: iot-edge-spiffe-agent/workload-api-server/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Error while trying the get trust bundle from server {0}")]
+    TrustBundleResponse(Box<dyn std::error::Error + Send>),
+    #[error("Error while trying to convert the trust jwkset to vec<u8> {0}")]
+    SerdeConvertToVec(serde_json::Error),
```
Comment (nyanzebra) [inline iot-edge-spiffe-agent/workload-api-server/src/error.rs:10]: I might be misremembering, but I think the serde error will already have this information of what the failed conversion was? Also, I think thiserror crate has a '#[from' macro that can be put in the enum variant to allow you to do err.into() :)

File: iot-edge-spiffe-server/key-manager/src/error.rs [hunk 4]
```diff
@@ -1,11 +1,18 @@
...
 #[derive(Error, Debug)]
 pub enum Error {
     #[error("Error while creating a new private key {0}")]
     CreatingNewKey(#[from] Box<dyn std::error::Error>),
+    #[error("Converting key to ec key {0}")]
+    ECkeyConvertion(ErrorStack),
+    #[error("Error creating big num object {0}")]
+    BigNumGeneration(ErrorStack),
+    #[error("Error while generating X and Y {0}")]
+    GeneratyXandY(ErrorStack),
```
Comment (nyanzebra) [inline iot-edge-spiffe-server/key-manager/src/error.rs:15]: nit: Generating?

**Review by nyanzebra — APPROVED** [review]
Just some minor comments

File: common/core-objects/src/lib.rs [hunk 5]
```diff
@@ -153,6 +153,25 @@ pub enum KeyType {
...
+            KeyType::ES512 => (Kty::EC, Crv::P521),
+            _ => unimplemented!(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Serialize, PartialEq)]
+pub enum KeyUse {
+    #[serde(rename = "x509-svid")]
+    X509XVID,
```
Comment (varunpuranik) [inline common/core-objects/src/lib.rs:170]: typo: X509SVID

File: common/workload-api/proto/workloadapi.proto [hunk 6]
```diff
@@ -0,0 +1,93 @@
+syntax = "proto3";
```
Comment (varunpuranik) [inline common/workload-api/proto/workloadapi.proto:1]: How do we keep this in sync with SPIFFE workload.proto https://github.com/spiffe/go-spiffe/blob/main/v2/proto/spiffe/workload/workload.proto
Can we not ref it instead?

Comment (huguesBouvier) [reply common/workload-api/proto/workloadapi.proto:1]: You're right, I kept it local for convenience when I am developing but it doesn't make sense to keep a local replica.

Comment (huguesBouvier) [reply common/workload-api/proto/workloadapi.proto:1]: Wanted to submodule one file but seems it is not possible. Instead I am wgetting the file, write it on disk, then parse it.

File: common/core-objects/src/lib.rs (hunk 5 above)
Comment (huguesBouvier) [reply common/core-objects/src/lib.rs:170]: Fixed

File: iot-edge-spiffe-server/key-manager/src/error.rs (hunk 4 above)
Comment (huguesBouvier) [reply iot-edge-spiffe-server/key-manager/src/error.rs:15]: Fixed

File: iot-edge-spiffe-agent/workload-api-server/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+        bundles_map.insert(trust_bundle.trust_domain, jwk_set);
+
+        let trust_bundle_response = JwtBundlesResponse {
+            bundles: bundles_map,
+        };
+
+        let stream: Self::FetchJWTBundlesStream = Box::pin(async_stream::stream! {
+                yield Ok(trust_bundle_response)
+        }) as Self::FetchJWTBundlesStream;
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply iot-edge-spiffe-agent/workload-api-server/src/lib.rs:83]: You're right, it works. Thanks!

File: iot-edge-spiffe-agent/workload-api-server/src/error.rs (hunk 3 above)
Comment (huguesBouvier) [reply iot-edge-spiffe-agent/workload-api-server/src/error.rs:10]: Ah yes, I am aware but in this specific case, I can't. The reason is that the ? need to return a tonic status and not an Error::error.

File: Cargo.toml (hunk 2 above)
Comment (huguesBouvier) [reply Cargo.toml:24]: Done, thanks

File: iot-edge-spiffe-agent/workload-api-server/src/unix_stream.rs (hunk 1 above)
Comment (huguesBouvier) [reply iot-edge-spiffe-agent/workload-api-server/src/unix_stream.rs:1]: Done

File: common/workload-api/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,470 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (varunpuranik) [inline common/workload-api/src/lib.rs:1]: Is this file entirely auto-generated except this 1st line?

File: iot-edge-spiffe-agent/workload-api-server/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        _request: Request<JwtsvidRequest>,
+    ) -> Result<Response<JwtsvidResponse>, tonic::Status> {
+        todo!()
+    }
+
+    async fn fetch_jwt_bundles(
+        &self,
+        _request: Request<JwtBundlesRequest>,
+    ) -> Result<Response<Self::FetchJWTBundlesStream>, tonic::Status> {
+        let mut bundles_map = HashMap::new();
```
Comment (varunpuranik) [inline iot-edge-spiffe-agent/workload-api-server/src/lib.rs:60]: Where is workload attestation handled?

**Review by varunpuranik — APPROVED** [review]

File: iot-edge-spiffe-agent/workload-api-server/src/lib.rs (hunk 9 above)
Comment (huguesBouvier) [reply iot-edge-spiffe-agent/workload-api-server/src/lib.rs:60]: Not there yet, I have it as a pbi this sprint:
image
We don't need authentication for trust bundle. That's the same for workload API in iotedge.

File: common/workload-api/src/lib.rs (hunk 8 above)
Comment (huguesBouvier) [reply common/workload-api/src/lib.rs:1]: Yes, everything except the first line. I can remove the header here, I just did that as an automatism.

Comment (huguesBouvier) [reply common/workload-api/src/lib.rs:1]: Actually I remember, it was to pass the automated checks. I could have excluded the file too.

Comment (varunpuranik) [reply common/workload-api/src/lib.rs:1]: No, I was just curious. We do add the header for other generated files as well, so this is fine. https://github.com/Azure/iotedge/blob/main/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/version_2018_06_28/generatedCode/EdgeletHttpClient.cs#L1

**Pull request merged**